src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs

```diff
@@ -0,0 +1,376 @@
+            }
+
+            return map;
+        }
+
+        public void OnCompilationEventGenerated(CompilationEvent compilationEvent)
+        {
+            // Add the event to our global pending event cache.
+            UpdateEventsCache(compilationEvent, generated: true);
+
+            // Mark the event for analysis for each analyzer.
+            foreach (var analyzerState in _analyzerStateMap.Values)
+            {
+                analyzerState.OnCompilationEventGenerated(compilationEvent);
+            }
+        }
+
+        private void AnalyzeSymbolEvent(SymbolDeclaredCompilationEvent symbolDeclaredEvent, IEnumerable<DiagnosticAnalyzer> analyzers)
+        {
+            foreach (var analyzer in analyzers)
+            {
+                _analyzerStateMap[analyzer].AnalyzeSymbolEvent(symbolDeclaredEvent);
+            }
+        }
+
+        /// <summary>
+        /// Invoke this method at completion of event processing for the given analysis scope.
+        /// It updates the analysis state of this event for each analyzer and if the event has been fully processed for all analyzers, then removes it from our event cache.
+        /// </summary>
+        public void AnalyzeEvent(CompilationEvent compilationEvent, AnalysisScope analysisScope)
+        {
+            // Analyze if the symbol and all its declaring syntax references are analyzed.
+            var symbolDeclaredEvent = compilationEvent as SymbolDeclaredCompilationEvent;
+            if (symbolDeclaredEvent != null)
+            {
+                AnalyzeSymbolEvent(symbolDeclaredEvent, analysisScope.Analyzers);
+            }
+
+            // Check if event is fully analyzed for all analyzers.
+            foreach (var analyzerState in _analyzerStateMap.Values)
+            {
+                if (!analyzerState.IsEventAnalyzed(compilationEvent))
+                {
+                    return;
+                }
+            }
+
+            // Remove the event from event cache.
+            UpdateEventsCache(compilationEvent, generated: false);
+        }
+
+        public void UpdateEventsCache(CompilationEvent compilationEvent, bool generated)
+        {
+            var symbolEvent = compilationEvent as SymbolDeclaredCompilationEvent;
+            if (symbolEvent != null)
+            {
+                // Add/remove symbol events to per-tree event cache.
+                // Any diagnostics request for a tree should trigger symbol and syntax node analysis for symbols with at least one declaring reference in the tree.
+                foreach (var location in symbolEvent.Symbol.Locations)
+                {
+                    if (location.SourceTree != null)
+                    {
+                        if (generated)
+                        {
+                            _eventsCache.AddOrUpdate(location.SourceTree, ImmutableHashSet.Create(compilationEvent), (tree, events) => events.Add(compilationEvent));
```

Comment:
if this set mutates a lot, then immutable collection might not a good one to use since it allocates a lot on each mutation.